lib/types.nix

```diff
@@ -126,6 +126,20 @@ rec {
       substSubModules = m: listOf (elemType.substSubModules m);
     };
 
+    # a listOf type with a restriction on valid values
+    # example: types.someOf types.package [ pkgs.vim pkgs.emacs ]
+    someOf = elemType: validValues:
+      let checkValues = validValues: values:
+        let invalidValues = subtractLists validValues values; 
+            showList = xs: "[${concatStringsSep " " (map toString xs)}]";
+        in
+        if (invalidValues == []) then true
+        else builtins.trace (''
```

Comment:
Type checks should not call `trace`. Consider for instance `either (someOf ...) (...)` - you don't want a trace message to appear if the left-hand side of `either` doesn't match but the right-hand side does.